File: core/python/astal/application.py
import sys

from collections.abc import Callable
from gi.repository import Astal, Gio

RequestHandler = Callable[[str, Callable[[str], None]], None]


class _Application(Astal.Application):
    __gtype_name__ = "AstalPython"

    def __init__(self) -> None:
        super().__init__()
        self.request_handler: RequestHandler | None = None

    def request(self, msg: str, conn: Gio.SocketConnection) -> None:
        if self.request_handler:
            self.request_handler(
                msg,
                lambda response: Astal.write_sock(
                    conn,
                    str(response),
                    lambda _, res: Astal.write_sock_finish(res),
                ),
            )
        else:
            super().request(msg, conn)

    def start(
        self,
        instance_name: str | None = None,
        gtk_theme: str | None = None,
        icon_theme: str | None = None,
        cursor_theme: str | None = None,
        css: str | None = None,
        hold: bool | None = True,
        request_handler: RequestHandler | None = None,
        main: Callable | None = None,
    ) -> None:
        if request_handler:
            self.request_handler = request_handler
        if hold:
            self.hold()
        if instance_name:
            self.set_instance_name(instance_name)
        if gtk_theme:
            self.set_gtk_theme(gtk_theme)
        if icon_theme:
            self.set_icon_theme(icon_theme)
        if cursor_theme:
            self.set_cursor_theme(cursor_theme)
        if css:
            self.apply_css(css, False)
        if not self.acquire_socket():
            return print(
                f"Astal instance {self.get_instance_name()} already running",
                file=sys.stderr,
            )

        def on_activate(_):
            if main:
                main()

        self.connect("activate", on_activate)
        self.run()


App = _Application()

File: core/lua/astal/application.lua
local lgi = require("lgi")
local Astal = lgi.require("Astal", "0.1")

local AstalLua = Astal.Application:derive("AstalLua")
local request_handler

function AstalLua:do_request(msg, conn)
    if type(request_handler) == "function" then
        request_handler(msg, function(response)
            Astal.write_sock(conn, tostring(response), function(_, res)
                Astal.write_sock_finish(res)
            end)
        end)
    else
        Astal.Application.do_request(self, msg, conn)
    end
end

function AstalLua:quit(code)
    Astal.Application.quit(self)
    os.exit(code)
end

local app = AstalLua()

---@class StartConfig
---@field icons? string
---@field instance_name? string
---@field gtk_theme? string
---@field icon_theme? string
---@field cursor_theme? string
---@field css? string
---@field hold? boolean
---@field request_handler? fun(msg: string, response: fun(res: any))
---@field main? fun(...): unknown
---@field client? fun(message: fun(msg: string): string, ...): unknown

---@param config StartConfig | nil
function Astal.Application:start(config)
    if config == nil then
        config = {}
    end

    if config.client == nil then
        config.client = function()
            print('Astal instance "' .. app.instance_name .. '" is already running')
            os.exit(1)
        end
    end

    if config.hold == nil then
        config.hold = true
    end

    request_handler = config.request_handler

    if config.css then
        self:apply_css(config.css)
    end
    if config.icons then
        self:add_icons(config.icons)
    end
    if config.instance_name then
        self.instance_name = config.instance_name
    end
    if config.gtk_theme then
        self.gtk_theme = config.gtk_theme
    end
    if config.icon_theme then
        self.icon_theme = config.icon_theme
    end
    if config.cursor_theme then
        self.cursor_theme = config.cursor_theme
    end

    app.on_activate = function()
        if type(config.main) == "function" then
            config.main(table.unpack(arg))
        end
        if config.hold then
            self:hold()
        end
    end

    if not app:acquire_socket() then
        return config.client(function(msg)
            return Astal.Application.send_message(self.instance_name, msg)
        end, table.unpack(arg))
    end

    self:run(nil)
end

return app
